libportal/notification.c

```diff
@@ -60,22 +73,312 @@ ensure_action_invoked_connection (XdpPortal *portal)
 }
 
 typedef struct {
-  XdpPortal *portal;
-  GAsyncReadyCallback callback;
-  gpointer data;
-} CallDoneData;
+  GUnixFDList *fd_list;
+  GVariantBuilder *builder;
+  gsize parse_ref;
+} ParserData;
+
+static ParserData*
+parser_data_new (const char *id)
+{
+  ParserData *data;
+
+  data = g_new0 (ParserData, 1);
+  data->fd_list = g_unix_fd_list_new ();
+  data->builder = g_variant_builder_new (G_VARIANT_TYPE ("(sa{sv})"));
+
+  g_variant_builder_add (data->builder, "s", id);
+  g_variant_builder_open (data->builder, G_VARIANT_TYPE ("a{sv}"));
+
+  return data;
+}
+
+static void
+parser_data_hold (ParserData *data)
+{
+  data->parse_ref++;
+}
+
+static gboolean
+parser_data_release (ParserData *data)
+{
+  data->parse_ref--;
+
+  if (data->parse_ref == 0)
+    g_variant_builder_close (data->builder);
+
+  return data->parse_ref == 0;
+}
```

Comment:
The common terminology in GLib land is `ref` and `unref`. You could also use `gatomicrefcount` for this.

Reply:
I used hold/release since it extends the lifetime of parser similar to `g_application_hold/release()` But i'm also fine with changing it

Reply:
i renamed the `parse_ref` to make clear that this isn't actually freeing anything.